File: Ass_3_5.py
import sys
from collections import defaultdict

def get_subset_sums(arr):
    """
    Generates all possible subset sums for a list of integers.
    """
    sums = [0]
    for num in arr:
        # Create new sums by adding the current number to all existing sums
        current_len = len(sums)
        for i in range(current_len):
            sums.append(sums[i] + num)
    return sums

def solve():
    """
    Main function to handle test cases and input/output.
    """
    # Read all input from stdin at once
    input_data = sys.stdin.read().split()
    
    if not input_data:
        return

    iterator = iter(input_data)
    
    try:
        # Get number of test cases
        t_str = next(iterator, None)
        if not t_str: return
        T = int(t_str)
        
        for _ in range(T):
            # Get N and S
            N = int(next(iterator))
            S = int(next(iterator))
            
            # Get the array elements
            A = []
            for _ in range(N):
                A.append(int(next(iterator)))
            
            # --- Meet-in-the-Middle Algorithm ---
            
            # 1. Divide the array into two halves
            mid = N // 2
            left_part = A[:mid]
            right_part = A[mid:]
            
            # 2. Generate all subset sums for both halves
            left_sums = get_subset_sums(left_part)
            right_sums = get_subset_sums(right_part)
            
            # 3. Count frequencies of sums in the right half
            right_freq = defaultdict(int)
            for val in right_sums:
                right_freq[val] += 1
            
            # 4. Check combinations
            # For every sum 'u' in left_sums, we need 'v' in right_sums such that u + v = S
            # So, we look for v = S - u
            ans = 0
            for u in left_sums:
                target = S - u
                if target in right_freq:
                    ans += right_freq[target]
            
            print(ans)
            
    except StopIteration:
        pass

if __name__ == "__main__":
    solve()

File: Ass_3_5.java

import java.util.*;

public class Ass_3_5 {

    // Helper function to generate all subset sums for a given array part
    private static ArrayList<Long> getSubsetSums(long[] arr) {
        ArrayList<Long> sums = new ArrayList<>();
        sums.add(0L); // Start with sum 0 (empty subset)

        for (long num : arr) {
            int currentSize = sums.size();
            // For every existing sum, add the current number to create new sums
            for (int i = 0; i < currentSize; i++) {
                sums.add(sums.get(i) + num);
            }
        }
        return sums;
    }

    public static void main(String[] args) {
        // Use Scanner for reading input
        Scanner sc = new Scanner(System.in);

        // Check if there is input available
        if (!sc.hasNext()) {
            return;
        }

        try {
            int T = sc.nextInt(); // Number of test cases

            while (T-- > 0) {
                int N = sc.nextInt();
                long S = sc.nextLong(); // Target sum can be large

                long[] A = new long[N];
                for (int i = 0; i < N; i++) {
                    A[i] = sc.nextLong();
                }

                // --- Meet-in-the-Middle Logic ---
                // 1. Split array into two halves
                int mid = N / 2;
                long[] leftPart = Arrays.copyOfRange(A, 0, mid);
                long[] rightPart = Arrays.copyOfRange(A, mid, N);

                // 2. Generate all subset sums for both halves
                ArrayList<Long> leftSums = getSubsetSums(leftPart);
                ArrayList<Long> rightSums = getSubsetSums(rightPart);

                // 3. Store right_sums frequency in a HashMap for O(1) lookup
                HashMap<Long, Integer> rightFreq = new HashMap<>();
                for (long val : rightSums) {
                    rightFreq.put(val, rightFreq.getOrDefault(val, 0) + 1);
                }

                // 4. Iterate through left_sums and check for complement in right_sums
                long count = 0;
                for (long u : leftSums) {
                    long target = S - u;
                    if (rightFreq.containsKey(target)) {
                        count += rightFreq.get(target);
                    }
                }

                System.out.println(count);
            }
        } catch (NoSuchElementException e) {
            // Handle cases where input might stop abruptly
        } finally {
            sc.close();
        }
    }
}
